test/controllers/tasks_controller_test.rb

```diff
@@ -59,70 +76,161 @@
         post tasks_path, params: task_hash
-      }.must_change "Task.count", 1
+      }.must_change 'Task.count', 1
 
       new_task = Task.find_by(name: task_hash[:task][:name])
       expect(new_task.description).must_equal task_hash[:task][:description]
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
-      expect(new_task.completed).must_equal task_hash[:task][:completed]
+      expect(new_task.completion_date).must_equal task_hash[:task][:completion_date]
 
       must_respond_with :redirect
-      must_redirect_to task_path(new_task.id)
+      must_redirect_to tasks_path
     end
   end
 
   # Unskip and complete these tests for Wave 3
-  describe "edit" do
-    it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+  describe 'edit' do
+    it 'can get the edit page for an existing task' do
+      test_task = task
+
+      get edit_task_path(task.id)
+
+      must_respond_with :ok
     end
 
-    it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+    it 'will respond with redirect when attempting to edit a nonexistant task' do
+      get edit_task_path(-1)
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
     end
   end
 
   # Uncomment and complete these tests for Wave 3
-  describe "update" do
+  describe 'update' do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
-    it "can update an existing task" do
-      skip
-      # Your code here
+    it 'can update an existing task' do
+      test_task = task
+      task_hash = {
+        task: {
+          name: 'new task',
+          description: 'new task description',
+          completion_date: nil,
+          completed: true
+        }
+      }
+
+      # Act-Assert
+      expect {
+        patch task_path(test_task), params: task_hash
+      }.wont_change 'Task.count'
+
+      must_respond_with :redirect
+      must_redirect_to task_path
     end
 
-    it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+    it 'will redirect to the root page if given an invalid id' do
+      # Act
+      patch task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
+      must_redirect_to root_path
+      
     end
   end
 
   # Complete these tests for Wave 4
-  describe "destroy" do
-    # Your tests go here
+  describe 'destroy' do
+    it 'removes the task from the database' do
+      # Arrange
+      test_task = task
 
+      # Act
+      expect {
+        delete task_path(test_task)
+      }.must_change 'Task.count', -1
+
+      # Assert
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+
+      after_task = Task.find_by(id: task.id)
+      expect(after_task).must_be_nil
+    end
+
+    it 'returns 404 for a task that does not exist' do
+      # Arrange
+      task_id = 123_456
```

Comment:
_minor issue_, this ID could theoretically be a valid ID.  Instead I suggest using something like -1 or "invalid id" as the id.